### PR TITLE
actor: add inihibtor FirewalldCheckAllowZoneDrifting

### PR DESCRIPTION
**Pull request by erig0**
This config option does not exist in RHEL-9. So we must make sure it's
not being used.

Fixes: OAMG-4391

### CODE REVIEW

**Comment by erig0** [comment]
review please

File: repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    AllowZoneDrifiting=no.
+    """
+
+    name = 'firewalld_check_allow_zone_drifting'
+    consumes = (FirewallsFacts,)
+    produces = (Report,)
+    tags = (ChecksPhaseTag, IPUWorkflowTag)
+
+    def process(self):
+        conf_key = "AllowZoneDrifting"
```
Comment (vinzenz) [inline repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:22]:
```suggestion
        conf_key = 'AllowZoneDrifting'
```

File: repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        conf_dict = private.read_config(path)
+        if private.isAllowZoneDriftingInUse(conf_dict):
+            create_report([
+                reporting.Title("Unsupported configuration"),
+                reporting.Summary("Firewalld is using configuration option "
+                                  + "'%s' which has been removed in RHEL-9. "
+                                  % (conf_key)
+                                  + "New behavior is as if '%s' was set to 'no'."
+                                  % (conf_key)),
```
Comment (vinzenz) [inline repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:48]: Please do not use % formatting, but use .format
So to get you an idea:
```python
'Starting with RHEL-9, FirewallD removed the support for the configuration option `{option}`.'.format(option=conf_key)
```

The variable '{option}' can be also used multiple times inside of the string, so no need to repeat the argument. That's the advantage of the named formatting strings.

**Review by vinzenz — CHANGES_REQUESTED** [review]
Thank you very much for your PR, it looks quite good already, however there are a few things in general that need to be addressed.

* Please drop the usage of double quotes and replace them with single quotes.
* Please do not use % style string formatting
* The detection part, is by our convention to be done in the FactsCollection phase (FactsPhaseTag)
    * The ChecksPhaseTag should check just for message provided by the fact collection actor and generate the inhibitor based on that. 
    *  For that you will have to create a Model that you can pass that information, e.g. to make it usable for the future, it could be a list of deprecated/removed options that are used by the user, etc.

File: repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/tests/component_test_firewalldcheckallowzonedrifting.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+from leapp.libraries.actor import private
```
Comment (fernflower) [inline repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/tests/component_test_firewalldcheckallowzonedrifting.py:1]: 'private' library name is to me a bit confusing. Maybe rename it to firewalldlib or just 'library' or something similar?

File: repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                reporting.Severity(reporting.Severity.HIGH),
+                reporting.Tags([reporting.Tags.SANITY]),
```
Comment (fernflower) [inline repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:50]: You can also add Tag.FIREWALL there

**Comment by erig0** [comment]
> Thank you very much for your PR, it looks quite good already, however there are a few things in general that need to be addressed.
> 
>     * Please drop the usage of double quotes and replace them with single quotes.

ACK. Both are valid and there is no PEP suggesting either. That being said, if Leapp uses single quotes everywhere I will change them.

>     * Please do not use % style string formatting

ACK. Bad habit from python < 2.7.

>     * The detection part, is by our convention to be done in the FactsCollection phase (FactsPhaseTag)
>       
>       * The ChecksPhaseTag should check just for message provided by the fact collection actor and generate the inhibitor based on that.
>       * For that you will have to create a Model that you can pass that information, e.g. to make it usable for the future, it could be a list of deprecated/removed options that are used by the user, etc.

Knowledge about `AllowZoneDrifting` is only useful to this actor. I based this code off a similar actor from el7_to_el8. I suppose I can rework it to add a Model.

File: repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/tests/component_test_firewalldcheckallowzonedrifting.py (hunk 3 above)
Comment (erig0) [reply repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/tests/component_test_firewalldcheckallowzonedrifting.py:1]: "private" is the convention used in the docs, but I'll change it.

File: repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py (hunk 1 above)
Comment (erig0) [reply repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:22]: For identifiers, that makes sense.

**Comment by erig0** [comment]
review please.

**Comment by Rezney** [comment]
@erig0 Thanks for PR. It is in a good shape now. I would just ask you to modify the report title a little. As for now it shows as:

```
Upgrade has been inhibited due to the following problems:
    1. Inhibitor: Unsupported configuration
Consult the pre-upgrade report for details and possible remediation.
```

I suggest to incorporate "firewalld" into the title.

**Comment by erig0** [comment]
Thanks. I folded and this change and pushed.

```python
diff --git a/repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py b/repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py
index 9f2a26a3706e..6db0b81c959d 100644
--- a/repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py
+++ b/repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py
@@ -33,7 +33,7 @@ class FirewalldCheckAllowZoneDrifting(Actor):
                 return
 
         create_report([
-            reporting.Title('Unsupported configuration'),
+            reporting.Title('Unsupported Firewalld Configuration'),
             reporting.Summary('Firewalld has enabled configuration option '
                               '"{conf_key}" which has been removed in RHEL-9. '
                               'New behavior is as if "{conf_key}" was set to "no".'.format(
```

**Comment by Rezney** [comment]
@erig0 Thank you. Please do a rebase and it should be good to go.

**Comment by erig0** [comment]
> @erig0 Thank you. Please do a rebase and it should be good to go.

Done.

**Comment by erig0** [comment]
> > @erig0 Thank you. Please do a rebase and it should be good to go.
> 
> Done.

bump

**Review by fernflower — APPROVED** [review]

File: repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+from leapp.reporting import create_report, Report
+from leapp.tags import ChecksPhaseTag, IPUWorkflowTag
+
+
+class FirewalldCheckAllowZoneDrifting(Actor):
+    '''
+    This actor will check if AllowZoneDrifiting=yes in firewalld.conf. This
+    option has been removed in RHEL-9 and behavior is as if
+    AllowZoneDrifiting=no.
+    '''
```
Comment (pirat89) [inline repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:13]: @erig0 pls use triple double quotes for all docstrings
* https://leapp.readthedocs.io/en/latest/python-coding-guidelines.html#write-docstrings-that-work-well-with-tools
* https://www.python.org/dev/peps/pep-0257/

Comment (erig0) [reply repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py]: done

File: repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                                  conf_key='AllowZoneDrifiting')),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Tags([reporting.Tags.SANITY, reporting.Tags.FIREWALL]),
+            reporting.Flags([reporting.Flags.INHIBITOR]),
+            reporting.ExternalLink(
+                url='https://access.redhat.com/articles/4855631',
+                title='Changes in firewalld related to Zone Drifting'),
+            reporting.Remediation(
+                hint='Set AllowZoneDrifting=no in /etc/firewalld/firewalld.conf',
+                commands=[['sh', '-c',
```
Comment (Rezney) [inline repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:49]: As this is just a single command, please remove the "sh -c" and let's go with the sed command only. 

Now, in the report, it translates to:
`sh -c sed -i "s/^AllowZoneDrifting=.*/AllowZoneDrifting=no/" /etc/firewalld/firewalld.conf` which is broken without apostrophes around the sed command.
Thanks...

Comment (erig0) [reply repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:49]: Oh. I thought this was doing shell escaping. I followed the `sh -c` pattern from this example: https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el7toel8/actors/checkhacluster/libraries/checkhacluster.py

I'll change it to just call `sed` directly.

Is there a way to test this via snactor?

Comment (Rezney) [reply repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:49]: I do not think we can test it via snactor but you can test it using Leapp cockpit plugin. I can share details with you offline.

Comment (Rezney) [reply repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py:49]: In any way, when the command is in a good shape and executable, we should be good to merge this.

Comment (erig0) [reply repos/system_upgrade/el8toel9/actors/firewalldcheckallowzonedrifting/actor.py]: done.

**Review by Rezney — APPROVED** [review]

**Comment by erig0** [comment]
@Rezney, thanks for working through these with me!

**Pull request merged**